nodeagent/nodeagentmanager.go

```diff
@@ -0,0 +1,340 @@
+		gsd.SessionCookie = sessionCookie
+		gsd.SessionID = sessionID
+		gsd.InitialPlayers = initialPlayers
+	}
+}
+
+func (n *NodeAgentManager) gameServerDeleted(objUnstructured interface{}) {
+	obj := objUnstructured.(*unstructured.Unstructured)
+	gameServerName := obj.GetName()
+	gameServerNamespace := obj.GetNamespace()
+
+	log.WithFields(log.Fields{
+		GameServerName:      gameServerName,
+		GameServerNamespace: gameServerNamespace,
+	}).Infof("GameServer %s/%s deleted", gameServerNamespace, gameServerName)
+
+	n.gameServerMap.Delete(gameServerName)
+}
+
+func (n *NodeAgentManager) heartbeatHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := context.Background()
+	re := regexp.MustCompile(`.*/v1/sessionHosts\/(.*?)(/heartbeats|$)`)
+	match := re.FindStringSubmatch(r.RequestURI)
+
+	gameServerName := match[1]
+
+	var hb HeartbeatRequest
+	err := json.NewDecoder(r.Body).Decode(&hb)
+	if err != nil {
+		badRequest(w, err, "cannot deserialize json")
+		return
+	}
+
+	if err := validateHeartbeatRequestArgs(&hb); err != nil {
+		badRequest(w, err, "invalid heartbeat request")
+		return
+	}
+
+	gsdi, exists := n.gameServerMap.Load(gameServerName)
+	if !exists {
+		// this will probably happen when the GameServer CR is deleted. Pod will probably send some heartbeats before its deleted
+		internalServerError(w, fmt.Errorf("game server %s not found", gameServerName), "gameserver not found")
+		return
+	}
+
+	gsd := gsdi.(*GameServerDetails)
+	logger := getLogger(gameServerName, gsd.GameServerNamespace)
+
+	if logEveryHeartbeat {
+		logger.Infof("heartbeat received from sessionHostId %s, data %#v", gameServerName, hb)
+	}
+
+	if err := n.updateHealthAndStateIfNeeded(ctx, &hb, gameServerName, gsd); err != nil {
+		logger.Errorf("error updating health %s", err.Error())
+		internalServerError(w, err, "error updating health")
+		return
+	}
+
+	if err := n.updateConnectedPlayersIfNeeded(ctx, &hb, gameServerName, gsd); err != nil {
+		logger.Errorf("error updating connected players count %s", err.Error())
+		internalServerError(w, err, "error updating connected players count")
+		return
+	}
+
+	var op GameOperation = GameOperationContinue
+
+	gsd.Mutex.RLock()
+	currentState := gsd.CurrentGameState
+	gsd.Mutex.RUnlock()
+
+	if currentState == GameStateInvalid { // user has not set the status yet
+		op = GameOperationContinue
+	} else if currentState == GameStateInitializing {
+		op = GameOperationContinue
+	} else if currentState == GameStateStandingBy {
+		op = GameOperationContinue
+	} else if currentState == GameStateActive {
+		op = GameOperationActive
+	} else if currentState == GameStateTerminated || currentState == GameStateTerminating {
+		op = GameOperationTerminate
+	}
+
+	sc := &SessionConfig{
+		SessionId:      gsd.SessionID,
+		SessionCookie:  gsd.SessionCookie,
+		InitialPlayers: gsd.InitialPlayers,
+	}
+
+	hr := &HeartbeatResponse{
+		Operation:     op,
+		SessionConfig: *sc,
+	}
+
+	json, _ := json.Marshal(hr)
```

Comment:
The error value is being ignored in this line. While unlikely we should be handling the case of a serialization error.